core/sched.cc

```diff
@@ -289,8 +290,26 @@ void cpu::reschedule_from_interrupt(bool called_from_yield,
             return;
 #endif
         } else if (!called_from_yield) {
-            auto &t = *runqueue.begin();
-            if (p->_runtime.get_local() < t._runtime.get_local()) {
+           auto &t = *runqueue.begin();
+            if (p->_realtime._priority > 0) {
+                // Only switch to a higher-priority realtime thread, or to
+                // the next one with equal priority if yielding.
+                if (t._realtime._priority < p->_realtime._priority) {
```

Comment:
This means that if somehow the current thread `p` has a high enough priority (for example 99) and all other threads in the run queue have a lower priority, this thread `p` will keep running forever, not even kernel threads will get a chance to do anything. Am I wrong in my analysis? Is this by design?

Reply:
Yes, it is by design that a real-time thread that never yields, is never preempted by any thread with a lower priority or non-real-time priority. 
However, the question about kernel threads is a good one, as in Linux/Unix the things that happen in the kernel weren't called "threads" and the priorities didn't apply to them. In particular cpu::load_balance() which can't even move the poor threads to other CPUs. 

I think the idea is that real-time threads should be short-running - and if a real-time thread just starts an endless loop that never returns, it will cause problems also on Linux.

Maybe the right thing to set some sort of special priority to cpu::load_balance (and anything else?) to allow it to run also when a real-time thread is running. I don't know. I think we can improve this later.